base/iterators.jl

```diff
@@ -749,9 +749,12 @@ julia> collect(Iterators.product(1:2, 3:5))
 """
 product(iters...) = ProductIterator(iters)
 
-IteratorSize(::Type{ProductIterator{Tuple{}}}) = HasShape{0}()
-IteratorSize(::Type{ProductIterator{T}}) where {T<:Tuple} =
-    prod_iteratorsize( IteratorSize(tuple_type_head(T)), IteratorSize(ProductIterator{tuple_type_tail(T)}) )
+function IteratorSize(::Type{ProductIterator{T}}) where {T<:Tuple}
+    if isvatuple(T)
+        throw(ArgumentError("Cannot compute IteratorSize for ProductIterator{$T}"))
+    end
+    return reduce(prod_iteratorsize, HasShape{0}(), ntuple(i -> IteratorSize(fieldtype(T, i))), fieldcount(T))
```

Comment:
Indeed, using `ntuple` and `reduce` in cases like this is pretty hard on the compiler. We should try to make tuple_type_tail work, or manually recur over field indices.